sdk/core/Azure.Core/tests/common/ModelReaderWriter/Models/DiscriminatorSet/BaseModel.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        private void Serialize(Utf8JsonWriter writer, ModelReaderWriterOptions options)
+        {
+            writer.WriteStartObject();
+            writer.WritePropertyName("kind"u8);
+            writer.WriteStringValue(Kind);
+            if (Optional.IsDefined(Name))
+            {
+                writer.WritePropertyName("name"u8);
+                writer.WriteStringValue(Name);
+            }
+            if (options.Format == ModelReaderWriterFormat.Json)
+            {
+                SerializeRawData(writer);
+            }
+            writer.WriteEndObject();
+        }
+
+        internal static BaseModel DeserializeBaseModel(BinaryData data, ModelReaderWriterOptions options)
+            => DeserializeBaseModel(JsonDocument.Parse(data.ToString()).RootElement, options);
+
+        internal static BaseModel DeserializeBaseModel(JsonElement element, ModelReaderWriterOptions options = default)
+        {
+            options ??= ModelReaderWriterOptions.DefaultWireOptions;
+
+            if (element.ValueKind == JsonValueKind.Null)
+            {
+                return null;
+            }
+            if (element.TryGetProperty("kind", out JsonElement discriminator))
+            {
+                switch (discriminator.GetString())
+                {
+                    case "X":
+                        return ModelX.DeserializeModelX(element, options);
+                    case "Y":
+                        return ModelY.DeserializeModelY(element, options);
+                }
+            }
+
+            //Deserialize unknown subtype
+            string kind = default;
+            Optional<string> name = default;
+            Dictionary<string, BinaryData> rawData = new Dictionary<string, BinaryData>();
+            foreach (var property in element.EnumerateObject())
+            {
+                if (property.NameEquals("kind"u8))
+                {
+                    kind = property.Value.GetString();
+                    continue;
+                }
+                if (property.NameEquals("name"u8))
+                {
+                    name = property.Value.GetString();
+                    continue;
+                }
+                if (options.Format == ModelReaderWriterFormat.Json)
+                {
+                    //this means it's an unknown property we got
+                    rawData.Add(property.Name, BinaryData.FromString(property.Value.GetRawText()));
+                }
+            }
+            return new UnknownBaseModel(kind, name, rawData);
+        }
+
+        BaseModel IModel<BaseModel>.Read(BinaryData data, ModelReaderWriterOptions options)
+        {
+            ModelSerializerHelper.ValidateFormat(this, options.Format);
+
+            return DeserializeBaseModel(JsonDocument.Parse(data.ToString()).RootElement, options);
+        }
+
+        BaseModel IJsonModel<BaseModel>.Read(ref Utf8JsonReader reader, ModelReaderWriterOptions options)
```

Comment:
If we put this in a real client without calling it through the `IJsonModel` interface, we trigger `AZC0014`.  What is our plan for addressing this?

The intent of this question is to answer the larger question: "we have `AZC0014` to prevent service clients from exposing types that lock us into implementation details when it comes to JSON reading/writing.  We can bypass the analyzer because of the way the analyzer is implemented, but we are still violating the intent of the analyzer in our client libraries now.  If we are ok exposing these implementation details, should we remove the analyzer?  Why or why not?"

@KrzysztofCwalina FYI

Reply:
it doesn't trigger azc0014 because its explicitly implemented.

Reply:
I know that :)

That's essentially working around an analyzer implementation though.  The intent of the analyzer is to prevent us from exposing JSON implementation details.

I am fairly confident the answer is that we are happy exposing those implementation details in this case, since we've spent a lot of time looking at those APIs, but I do expect customers to sometimes not know the need to explicitly implement the interface and run into AZC0014.  So, my question is - are we ok if they suppress the warning?  Will we catch it in PR reviews?  Do we care about AZC0014 anymore at all and should we remove it?

Reply:
I don't think anything about AZC0014 needs to change.  If we want to change it I think that is a discussion we should have outside the scope of this PR.  The use of Utf8JsonReader and Utf8JsonWriter are intended in the public API surface

Reply:
Can you please answer the remaining questions about how we will message this to client authors that make mistakes?

Reply:
Add azure sdk client analyzers back